### PR TITLE
Tutorial 2: custom constraint

### PR DESCRIPTION
**Pull request by jxmorris12**

### CODE REVIEW

File: examples/[2] Constraints.ipynb [hunk 1]
```diff
@@ -0,0 +1,508 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# The importance of constraints\n",
+    "\n",
+    "Constraints determine which potential adversarial examples are valid inputs to the model. When determining the efficacy of an attack, constraints are everything. After all, an attack that looks very powerful may just be generating nonsense. That's why you should always clearly define the *constraints* your adversarial examples must meet. \n",
```
[the remark this answers is not in the document]
Comment (jxmorris12) [reply examples/[2] Constraints.ipynb:9]: thanks, tried to clarify this.

File: examples/[2] Constraints.ipynb [hunk 2]
```diff
@@ -0,0 +1,508 @@
...
+    "\n",
+    "### Classes of constraints\n",
+    "\n",
+    "TextAttack evaluates constraints using methods from three groups:\n",
+    "\n",
+    "- **Overlap constraints** determine if a perturbation is valid based on character-level analysis. For example, some attacks are constrained by edit distance: a perturbation is only valid if it perturbs some small number of characters (or fewer).\n",
+    "\n",
+    "- **Syntactic constraints** filter inputs based on their syntax. For example, an attack may require that adversarial examples do not introduce grammatical errors.\n",
```
[the remark this answers is not in the document]
Comment (jxmorris12) [reply examples/[2] Constraints.ipynb:17]: I used perturbation for both

File: examples/[2] Constraints.ipynb [hunk 3]
```diff
@@ -0,0 +1,508 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## A custom constraint\n",
+    "\n",
+    "\n",
+    "For fun, we're going to see what happens when we constrain an attack to only allow perturbations that substitute out a named entity for another. In linguistics, a **named entity** is a proper noun, the name of a person, organization, location, product, etc. Named Entity Recognition is a popular NLP task (and one that state-of-the-art models can perform quite well). \n",
```
[the remark this answers is not in the document]
Comment (jxmorris12) [reply examples/[2] Constraints.ipynb:41]: hahaha yeah. I remember that too. They crafted them manually though!

File: examples/[2] Constraints.ipynb [hunk 4]
```diff
@@ -0,0 +1,508 @@
...
+    "\n",
+    "\n",
+    "### NLTK and Named Entity Recognition\n",
+    "\n",
+    "**NLTK**, the Natural Language Toolkit, is a Python package that helps developers write programs that process natural language. NLTK comes with predefined algorithms for lots of linguistic tasks– including Named Entity Recognition.\n",
```
[the remark this answers is not in the document]
Comment (jxmorris12) [reply examples/[2] Constraints.ipynb:46]: I want to make a tutorial specifically about sentence encoders - will add to Asana

File: examples/[2] Constraints.ipynb [hunk 5]
```diff
@@ -0,0 +1,508 @@
...
+   "source": [
+    "# The importance of constraints\n",
+    "\n",
+    "Constraints determine which potential adversarial examples are valid inputs to the model. When determining the efficacy of an attack, constraints are everything. After all, an attack that looks very powerful may just be generating nonsense. That's why you should always clearly define the *constraints* your adversarial examples must meet. \n",
+    "\n",
+    "### Classes of constraints\n",
+    "\n",
+    "TextAttack evaluates constraints using methods from three groups:\n",
+    "\n",
+    "- **Overlap constraints** determine if a perturbation is valid based on character-level analysis. For example, some attacks are constrained by edit distance: a perturbation is only valid if it perturbs some small number of characters (or fewer).\n",
```
[the remark this answers is not in the document]
Comment (jxmorris12) [reply examples/[2] Constraints.ipynb:15]: I think words still counts as character level - I'm adding task to asana about the renames

File: examples/[2] Constraints.ipynb [hunk 6]
```diff
@@ -0,0 +1,508 @@
...
+    "            # If we find two words that don't match, this means a word was swapped \n",
+    "            # between `x` and `x_adv`. That word must be a named entity to fulfill our\n",
+    "            # constraint.\n",
+    "            x_word_label = None\n",
+    "            x_adv_word_label = None\n",
+    "            for (word_1, label_1), (word_2, label_2) in zip(x_entities, x_adv_entities):\n",
+    "                if word_1 != word_2:\n",
+    "                    # Finally, make sure that words swapped between `x` and `x_adv` are named entities. If \n",
+    "                    # they're not, then we also return False.\n",
+    "                    if (label_1 not in ['NNP', 'NE']):\n",
```
[the remark this answers is not in the document]
Comment (jxmorris12) [reply examples/[2] Constraints.ipynb:263]: Good catch!! I think this is just because I put a checkpoint in the PR that didn't have the `or` in the if statement on line 263. I re-saved and it's in now.

File: examples/[2] Constraints.ipynb [hunk 7]
```diff
@@ -0,0 +1,508 @@
...
+   "source": [
+    "import torch\n",
+    "torch.cuda.is_available()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now, let's use our attack. We're going to iterate through the `YelpSentiment` dataset and attack samples until we achieve 10 successes. (There's a lot to check here, and since we're using beam search, each sample will take a few minutes. This will take a few hours to run on a single core.)"
```
[the remark this answers is not in the document]
Comment (jxmorris12) [reply examples/[2] Constraints.ipynb]: lol. We were until it took too long and I changed it to greedy.

[the remark this answers is not in the document]
Comment (jxmorris12) [reply examples/[2] Constraints.ipynb]: thanks for noticing, I changed the text.

File: examples/[2] Constraints.ipynb [hunk 8]
```diff
@@ -0,0 +1,508 @@
...
+   "source": [
+    "sentence = 'Jack Black starred in the 2003 film classic \"School of Rock\".'\n",
+    "get_entities(sentence)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We flattened the tree of entities, so the return format is a list of `(word, entity type)` tuples. Looks like we have three labels here: 'Jack' and 'Black' labeled as 'PERSON', 'School' as a proper noun, and 'Rock' as a 'GPE' (geo-political entitity: it thinks Rock is the name of a place, a city or something). Whatever technique NLTK uses for named entity recognition may be a bit rough, but it did a pretty decent job here!"
```
[the remark this answers is not in the document]
Comment (jxmorris12) [reply examples/[2] Constraints.ipynb]: Ah yes also a good catch. Thanks. I changed the helper method to say `binary=True` which is why they're all just NNP (it made writing the code later on a bit easier). I edited the description. Too bad we lost this bit of precision, though, I thought the more specific labels were interesting.
